vertical-pod-autoscaler/pkg/utils/metrics/quality/quality.go

```diff
@@ -67,3 +112,59 @@ func ObserveUsageRecommendationRelativeDiff(usage, recommendation float64, isOOM
 func ObserveMissingRecommendation(isOOM bool, resource corev1.ResourceName) {
 	usageMissingRecommendationCounter.WithLabelValues(string(resource), strconv.FormatBool(isOOM)).Inc()
 }
+
+// ObserveUsageRecommendationDiff records absolute diff between usage and
+// recommendation.
+func ObserveUsageRecommendationDiff(usage, recommendation float64, isOOM bool, resource corev1.ResourceName) {
+	switch resource {
+	case corev1.ResourceCPU:
+		cpuUsageRecommendationDiff.Observe((usage - recommendation))
+	case corev1.ResourceMemory:
+		memoryUsageRecommendationDiff.WithLabelValues(strconv.FormatBool(isOOM)).Observe(usage - recommendation)
+	default:
+		klog.Warningf("Unknown resource: %v", resource)
+	}
+}
+
+// ObserveRecommendation records the value of recommendation.
+func ObserveRecommendation(recommendation float64, isOOM bool, resource corev1.ResourceName) {
+	switch resource {
+	case corev1.ResourceCPU:
+		cpuRecommendations.Observe(recommendation)
+	case corev1.ResourceMemory:
+		memoryRecommendations.WithLabelValues(strconv.FormatBool(isOOM)).Observe(recommendation)
+	default:
+		klog.Warningf("Unknown resource: %v", resource)
+	}
+}
+
+// ObserveQualityMetrics records all quality metrics that we can derive from usage and recommendation.
+func ObserveQualityMetrics(usage, recommendation float64, isOOM bool, resource corev1.ResourceName) {
+	ObserveRecommendation(recommendation, isOOM, resource)
+	ObserveUsageRecommendationDiff(usage, recommendation, isOOM, resource)
+	ObserveUsageRecommendationRelativeDiff(usage, recommendation, isOOM, resource)
+}
+
+// ObserveQualityMetricsRecommendationMissing records all quality metrics that we can derive from usage when recommendation is missing.
+func ObserveQualityMetricsRecommendationMissing(usage float64, isOOM bool, resource corev1.ResourceName) {
+	ObserveMissingRecommendation(isOOM, resource)
+	ObserveUsageRecommendationDiff(usage, 0, isOOM, resource)
+}
+
+func prependNegativeBuckets(buckets []float64) []float64 {
+	bucketCount := len(buckets)
+	result := make([]float64, bucketCount*2+1)
+	result[bucketCount] = 0
+	for i, bucket := range buckets {
+		if bucket < 0 {
+			panic(fmt.Sprintf("Positive buckets expected. Got: %v", bucket))
+		}
+		result[bucketCount-i-1] = -bucket
+		result[bucketCount+i+1] = bucket
+	}
+	return result
+}
+
+func getExponentialBuckets(min, max, step float64) []float64 {
```

Comment:
Upon reading this I'm not sure if the function together with comments at var definition bring value.
Maybe it would more straightforward to use prometheus function directly and only comment on what the exponential factor is in each case.
WDYT?

Reply:
Done.